Rename the secondary commit queue from just q to txCommitQFinal

ptschip opened the pull request:
This helps to clarify the purpose of the queue. Aso we make the queue
a global structure with it's own lock so that we can clear the queue
out if needed.

Resubmit the transactions from txCommitQFinal when doing a re-org and
also clear it afterwards. Must maintain good locking order with mempooo.cs
which prevent us from accidentally added txns to the mempool during a reorg.

ptschip committed:
Rename the secondary commit queue from just q to txCommitQFinal

This helps to clarify the purpose of the queue. Aso we make the queue
a global structure with it's own lock so that we can clear the queue
out if needed.

Resubmit the transactions from txCommitQFinal when doing a re-org and
also clear it afterwards. Must maintain good locking order with mempooo.cs
which prevent us from accidentally added txns to the mempool during a reorg.

ptschip committed:
Fix potential deadlock

Lock order should be cs_main -> mempool.cs -> txCommitQ -> txCommitQFinal
however we can end up deadlocking when we do SyncWithWallets() so to
get around that we can use a temporary queue pointer just for the step
when we sync the wallet while still keeping our locks in order so that
we don't have issues during a reorg.

gandrewstone commented on src/txmempool.h:
```diff
@@ -542,7 +542,7 @@ class CTxMemPool
     void forEachThenClear(const Lambda &f)
     {
-        WRITELOCK(cs);
+        AssertWriteLockHeld(cs);
```
the purpose of the forEachThenClear function was to bury the lock handling.  mapTx is a public variable so if a set of operations are needed at a higher level, we can take the lock and iterate manually. Make an _ version of the function...

ptschip replied:
done

ptschip committed:
Restore the previous version of forEachThenClear

and create an unlocked version _forEachThenClear which assumes
mempool.cs is already locked.
